tooling/ast_fuzzer/src/program/mod.rs

```diff
@@ -189,23 +233,40 @@ impl Context {
     fn gen_functions(&mut self, u: &mut Unstructured) -> arbitrary::Result<()> {
         let ids = self.function_declarations.keys().cloned().collect::<Vec<_>>();
         for id in ids {
-            let body = FunctionContext::new(self, id).gen_body(u)?;
-            let decl = self.function_decl(id);
-            let func = Function {
-                id,
-                name: decl.name.clone(),
-                parameters: decl.params.clone(),
-                body,
-                return_type: decl.return_type.clone(),
-                unconstrained: decl.unconstrained,
-                inline_type: decl.inline_type,
-                func_sig: decl.signature(),
-            };
-            self.functions.insert(id, func);
+            self.gen_function(u, id)?;
         }
         Ok(())
     }
 
+    /// Generate random function body.
+    fn gen_function(&mut self, u: &mut Unstructured, id: FuncId) -> arbitrary::Result<()> {
+        self.gen_function_with_body(u, id, |u, fctx| fctx.gen_body(u))
+    }
+
+    /// Generate function with a specified body generator.
+    fn gen_function_with_body(
+        &mut self,
+        u: &mut Unstructured,
+        id: FuncId,
+        f: impl Fn(&mut Unstructured, &mut FunctionContext) -> arbitrary::Result<Expression>,
+    ) -> arbitrary::Result<()> {
+        let mut fctx = FunctionContext::new(self, id);
+        let body = f(u, &mut fctx)?;
```

Comment:
Right, because of this. Maybe it could be `f: impl Fn(&mut Unstructured, FunctionContext)` instead?